platforms/ide/problems-api/src/main/java/org/gradle/api/problems/IdFactory.java

```diff
@@ -0,0 +1,83 @@
+/*
+ * Copyright 2024 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.api.problems;
+
+import org.gradle.api.Incubating;
+import org.gradle.api.problems.internal.DefaultIdFactory;
+
+/**
+ * A factory for creating problem ids and problem groups.
+ * <p>
+ * Clients should use the {@link #instance()} method to obtain an instance of this factory.
+ * <p>
+ * Custom implementation for this class is not supported.
+ *
+ * @since 8.13
+ */
+@Incubating
+public abstract class IdFactory {
+
+    /**
+     * Constructor.
+     *
+     * @since 8.13
+     */
+    protected IdFactory() {
+    }
+
+    /**
+     * Returns the default instance of the factory.
+     *
+     * @return the instance
+     * @since 8.13
+     */
+    public static IdFactory instance() {
```

Comment:
💭 Should this be a service?

Reply:
That's the interesting bit in the review. My stance is that problem Ids and groups are lightweight descriptors that Gradle and third party plugins both should be able to define as static data. ie: 

```
class MyProblemIds {
    public static final ProblemGroup  group = ...
    public static final ProblemId id1 = ...
    public static final ProblemId id2 = ...
}
```

I played around with making it a service but it made the the whole Problems API usage a fair bit clunky to use.